### PR TITLE
improve stringify of custom types

### LINKED ISSUE #189: Working with fp-ts data types should be as smooth as possible integrating with the rest of JS
fp-ts types (especially Collections) should generally be representable in JSON and strings, without error or being useless.
Bad cases:
- `chunk.toString()` `[object Object]`
- `JSON.stringify(chunk)`:
   ```
   TypeError: Converting circular structure to JSON
    --> starting at object with constructor 'ChunkImpl'
    --- property 'left' closes the circle
    at JSON.stringify (<anonymous>)
   ```
Great built-in examples:
- Array
- Date

Okay built-in examples:
- Object (JSON good, string bad)
- Map (JSON bad `{}`, string `[object Map]` - at least mentions Map)
  - this is at least an interesting built-in case, as it ends up as `{}` in the JSON, I suppose hiding internals might be the least step we should do, when not using another reasonable representation. And having `Map` in the name instead of `Object` is also better.

# The Chunk problem

Chunk is the most important imo, because it is being used (with good reason) in many places instead of Array, especially in Effect, Config, Cause, errors etc. Leading to problems displaying as string, or dumping to a basic JSON representation.

Of course Encoding and Parsing is not the point of this endeavour, that's what Schema is for. Special rendering is of course left to Show-like's, Printers etc.
But a basic, out of the box useful representation, especially for types that mirror or are extensions/replacements to more basic built-in types like Array, should provide the least friction possible.

My proposal is: Best Effort, basic useful integration with standard JS tools. We all know they have limitations, but that is accepted and out of scope of this proposal.

Example: https://github.com/patroza/data/pull/1/files
My thoughts on string representation are:
- `Chunk(1, 2, 3)`
- `Chunk<1, 2, 3>`
- `Chunk[1, 2, 3]`

When happy/aligned with the proposal, I will proceed with other data types.

## Related

Another topic is the way rendering in `console.log` works, in NodeJS this is done via `inspect` and there is a symbol you can override. Sadly this is not used in the browser console. And I know for instance @mikearnaldi is of opinion we should not specialise for any particular JS runtimes.
My opinion is slightly milder; as long as it doesn't cause any breakage, then improving the integration in some runtimes is a Nobel cause. However it means the behaviour is different across runtimes.
IMO that is a fact in general, not just limited to fp-ts/effect-ts libraries. I'm not sure that should stop us from at least making it better in some (major) runtimes.

Anyway, to me this is not as important as `toString()`and `toJSON()`, which are runtime agnostic.

- https://nodejs.org/api/util.html#util_util_inspect_custom
- https://www.npmjs.com/package/symbol.inspect

### PR DESCRIPTION
**Pull request by mikearnaldi**
closes #189

### CODE REVIEW

**Comment by mikearnaldi** [comment]
@patroza please review

File: src/Chunk.ts [hunk 1]
```diff
@@ -223,6 +223,21 @@ class ChunkImpl<A> implements Chunk<A> {
     }
   }
 
+  toString() {
+    return `Chunk(${this.toReadonlyArray().map(String).join(", ")})`
```
Comment (patroza) [inline src/Chunk.ts:227]: Is there a reason for the explicit `map(String).join(", ")` everywhere, over leveraging the built in Array toString?
`${this.toReadonlyArray()}`, then it's aligned with built-in

File: src/Chunk.ts [hunk 2]
```diff
@@ -223,6 +223,21 @@ class ChunkImpl<A> implements Chunk<A> {
...
+  }
+
+  toJSON() {
+    return {
+      _tag: "Chunk",
+      array: this.toReadonlyArray()
```
Comment (patroza) [inline src/Chunk.ts:233]: The other `toJSON` implementations seem to use `values` instead of `array`

**Review by patroza — COMMENTED** [review]
@mikearnaldi This is great, thanks! Left a few notes.

File: src/Chunk.ts (hunk 1 above)
Comment (mikearnaldi) [reply src/Chunk.ts:227]: The reason is `Chunk(0, 1, 2)` vs `Chunk(0,1,2)` which I like more

File: src/Chunk.ts (hunk 2 above)
Comment (mikearnaldi) [reply src/Chunk.ts]: good spot, resolved

**Pull request merged**